samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -230,22 +249,44 @@ public String getNextOpId(OpCode opCode) {
    * @param <M> the type of messages in the intermediate {@link MessageStream}
    * @return  the intermediate {@link MessageStreamImpl}
    */
-  @VisibleForTesting
-  <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
+  public <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
     Preconditions.checkState(!inputOperators.containsKey(streamId) && !outputStreams.containsKey(streamId),
         "getIntermediateStream must not be called multiple times with the same streamId: " + streamId);
 
-    if (serde == null) {
-      LOGGER.info("Using default serde for intermediate stream: " + streamId);
-      serde = (Serde<M>) defaultSerde;
+    Serde streamSerde;
+    if (serde != null) {
+      streamSerde = serde;
+    } else {
+      if (defaultSystemDescriptor != null) {
+        LOGGER.info("Using default system's serde for intermediate stream: " + streamId);
+        streamSerde = (Serde) defaultSystemDescriptor.getSystemSerde().orElse(null);
+      } else {
+        LOGGER.warn("No serde defined for intermediate stream: " + streamId +
+            ". Key and message serdes specified for the job.default.system will be used.");
+        streamSerde = null;
+      }
     }
 
     if (isBroadcast) broadcastStreams.add(streamId);
-    boolean isKeyed = serde instanceof KVSerde;
-    KV<Serde, Serde> kvSerdes = getKVSerdes(streamId, serde);
+
+    boolean isKeyed;
+    KV<Serde, Serde> kvSerdes;
+    if (streamSerde == null) { // if no explicit or default system descriptor serde available
+      isKeyed = true; // assume keyed stream
+      kvSerdes = new KV<>(null, null); // and that key and msg serdes are provided for job.default.system in configs
+    } else {
+      isKeyed = streamSerde instanceof KVSerde;
+      kvSerdes = getKVSerdes(streamId, streamSerde);
+    }
+
+    InputTransformer<?> transformer = null;
+    if (defaultSystemDescriptor instanceof TransformingSystemDescriptor) {
+      transformer = ((TransformingSystemDescriptor) defaultSystemDescriptor).getTransformer();
+    }
+
     InputOperatorSpec inputOperatorSpec =
         OperatorSpecs.createInputOperatorSpec(streamId, kvSerdes.getKey(), kvSerdes.getValue(),
-            isKeyed, this.getNextOpId(OpCode.INPUT, null));
+            transformer, isKeyed, this.getNextOpId(OpCode.INPUT, null));
```

Comment:
Could you pass the input descriptor as a container instead of extracting the properties individually?

Reply:
Thanks for the explanation. My original thought was that it would be a nice container for properties, to avoid updating the method signature with new args in the future, but it sounds like it's not actually worth it.